api4/plugin.go

```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
+// See License.txt for license information.
+
+// EXPERIMENTAL - SUBJECT TO CHANGE
+
+package api4
+
+import (
+	"net/http"
+
+	l4g "github.com/alecthomas/log4go"
+	"github.com/mattermost/platform/app"
+	"github.com/mattermost/platform/model"
+	"github.com/mattermost/platform/plugin"
+	"github.com/mattermost/platform/utils"
+)
+
+const (
+	MAXIMUM_PLUGIN_FILE_SIZE = 52428800 // 50 MB
+)
+
+func InitPlugin() {
+	l4g.Debug("EXPERIMENTAL: Initializing plugin api")
+
+	BaseRoutes.Plugins.Handle("", ApiSessionRequired(uploadPlugin)).Methods("POST")
+	BaseRoutes.Plugins.Handle("", ApiSessionRequired(getPlugins)).Methods("GET")
+	BaseRoutes.Plugin.Handle("", ApiSessionRequired(removePlugin)).Methods("DELETE")
+
+}
+
+func uploadPlugin(c *Context, w http.ResponseWriter, r *http.Request) {
+	if !*utils.Cfg.PluginSettings.Enable {
+		c.Err = model.NewAppError("uploadPlugin", "app.plugin.disabled.app_error", nil, "", http.StatusNotImplemented)
+		return
+	}
+
+	if !app.SessionHasPermissionTo(c.Session, model.PERMISSION_MANAGE_SYSTEM) {
+		c.SetPermissionError(model.PERMISSION_MANAGE_SYSTEM)
+		return
+	}
+
+	if err := r.ParseMultipartForm(MAXIMUM_PLUGIN_FILE_SIZE); err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	m := r.MultipartForm
+
+	pluginArray, ok := m.File["plugin"]
+	if !ok {
+		c.Err = model.NewAppError("uploadPlugin", "api.plugin.upload.no_file.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	if len(pluginArray) <= 0 {
+		c.Err = model.NewAppError("uploadPlugin", "api.plugin.upload.array.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+
+	file, err := pluginArray[0].Open()
+	if err != nil {
+		c.Err = model.NewAppError("uploadPlugin", "api.plugin.upload.file.app_error", nil, "", http.StatusBadRequest)
+		return
+	}
+	defer file.Close()
+
+	manifest, unpackErr := app.UnpackAndActivatePlugin(file)
```

Comment:
Got any plans for HA uploads?

Reply:
For the proof-of-concept it will be `upload plugins to all HA servers manually`, but we should come up with a more elegant solution

Reply:
I imagine we'll probably want to use the file driver like we do for attachments: When plugins are uploaded, upload them via that. Then when the plugin is activated, make a copy from the file driver to the local filesystem before running it.

Reply:
what if you upload a different plugin with the same name, does it gets overwritten? with that in mind, aren't we going to include a way to update existing plugins?

Reply:
If you upload one with the same name/id it will get overwritten as is. We don't have much handling for this case since we're still experimental